File: SkateboardAS/Infrastructure.Assembly/AssemblyService.cs
using System.Composition;
using Core.Attributes;
using Core.Enums;
using Core.Interfaces;
using Core.Lifecycle;
using Core.Models;
using Infrastructure.Assembly.Internal;

namespace Infrastructure.Assembly;

[Component("Assembly Service", "1.0.0")]
[Provides(typeof(IAssemblyService))]
[Provides(typeof(IMachineComponent))]
[Export(typeof(IAssemblyService))]
[Export(typeof(IMachineComponent))]
[ExportMetadata("Name", "Assembly Service")]
[ExportMetadata("Version", "1.0.0")]
[ExportMetadata("Protocol", "MQTT")]
[ExportMetadata("MachineType", "AssemblyStation")]
[ExportMetadata("Description", "UR assembly station for product assembly")]
[ExportMetadata("Icon", "cog")]
[ExportMetadata("Priority", 2)]
public class AssemblyService : IAssemblyService, IMachineComponent, IComponent
{
    private readonly MqttClientWrapper _mqttClient = new();

    public ComponentState State { get; private set; } = ComponentState.Installed;
    public string Id => "assembly-01";
    public string Name => "Assembly Station";
    public MachineType Type => MachineType.AssemblyStation;
    public MachineStatus Status { get; private set; } = MachineStatus.Offline;

    async Task IComponent.StartAsync(CancellationToken ct)
    {
        State = ComponentState.Starting;
        await _mqttClient.ConnectAsync(ct);
        Status = MachineStatus.Idle;
        State = ComponentState.Active;
    }

    async Task IComponent.StopAsync(CancellationToken ct)
    {
        State = ComponentState.Stopping;
        await _mqttClient.DisconnectAsync(ct);
        Status = MachineStatus.Offline;
        State = ComponentState.Uninstalled;
    }

    public Task<CommandResult> StartAsync(CancellationToken ct = default) => Task.FromResult(CommandResult.Ok());
    public Task<CommandResult> StopAsync(CancellationToken ct = default) => Task.FromResult(CommandResult.Ok());
    public Task<CommandResult> ResetAsync(CancellationToken ct = default) => Task.FromResult(CommandResult.Ok());
    public Task<MachineStatusModel> GetStatusAsync(CancellationToken ct = default)
        => Task.FromResult(new MachineStatusModel { Id = Id, Name = Name, State = Status.ToString() });

    public Task<CommandResult> StartOperationAsync(string processId, CancellationToken ct = default)
        => _mqttClient.PublishOperationAsync(processId, ct);

    public Task<bool> CheckHealthAsync(CancellationToken ct = default)
        => _mqttClient.CheckHealthAsync(ct);

    public Task<AssemblyStatus> GetAssemblyStatusAsync(CancellationToken ct = default)
        => _mqttClient.GetStatusAsync(ct);
}

File: SkateboardAS/infrastructure-assembly/src/main/java/dk/sdu/infrastructure/assembly/AssemblyService.java
package dk.sdu.infrastructure.assembly;

import dk.sdu.core.api.IAssemblyService;
import dk.sdu.core.api.IMachineComponent;
import dk.sdu.core.enums.MachineStatus;
import dk.sdu.core.enums.MachineType;
import dk.sdu.core.lifecycle.ComponentState;
import dk.sdu.core.lifecycle.MachineLifecycle;
import dk.sdu.core.metadata.ComponentMetadata;
import dk.sdu.core.model.AssemblyStatus;
import dk.sdu.core.model.CommandResult;
import dk.sdu.core.model.MachineStatusModel;

@ComponentMetadata(
    name = "Assembly Station",
    protocol = "MQTT",
    machineType = MachineType.ASSEMBLY_STATION,
    description = "Assembly station (UR) for part assembly",
    icon = "cog",
    priority = 0
)
public class AssemblyService implements IAssemblyService, IMachineComponent, MachineLifecycle {

    // IAssemblyService
    @Override
    public CommandResult startOperation(String processId) {
        // TODO: Implement
        return null;
    }

    @Override
    public boolean checkHealth() {
        // TODO: Implement
        return false;
    }

    @Override
    public AssemblyStatus getAssemblyStatus() {
        // TODO: Implement
        return null;
    }

    // IMachineComponent
    @Override
    public String getId() {
        // TODO: Implement
        return null;
    }

    @Override
    public String getName() {
        // TODO: Implement
        return null;
    }

    @Override
    public MachineType getType() {
        // TODO: Implement
        return null;
    }

    @Override
    public MachineStatus getStatus() {
        // TODO: Implement
        return null;
    }

    @Override
    public CommandResult start() {
        // TODO: Implement
        return null;
    }

    @Override
    public CommandResult stop() {
        // TODO: Implement
        return null;
    }

    @Override
    public CommandResult reset() {
        // TODO: Implement
        return null;
    }

    @Override
    public MachineStatusModel getStatusDetails() {
        // TODO: Implement
        return null;
    }

    // MachineLifecycle
    @Override
    public ComponentState getState() {
        // TODO: Implement
        return null;
    }

    @Override
    public void initialize() throws Exception {
        // TODO: Implement
    }

    @Override
    public void shutdown() throws Exception {
        // TODO: Implement
    }
}
